controller/src/main.cpp

```diff
@@ -98,16 +98,58 @@ static void DEV_MODE_comms_handler(const ControllerStatus &controller_status,
 }
 #endif
 
-static void controller_loop() {
-  Sensors sensors;
-  Controller controller;
+static Controller controller;
+static ControllerStatus controller_status;
+static Sensors sensors;
+
+// This function handles all the high priority tasks which need to be called
+// periodically.  The HAL calls this function from a timer interrupt.
+//
+// NOTE - it's important that anything being called from this function executes
+// quickly.  No busy waiting here.
+static void high_priority_task() {
+
+  // Read the sensors
+  controller_status.sensor_readings = sensors.GetSensorReadings();
+
+  // Run our PID loop
+  ActuatorsState actuators_state =
+      controller.Run(Hal.now(), controller_status.active_params,
+                     controller_status.sensor_readings);
+
+  // TODO update pb library to replace fan_power in ControllerStatus with
+  // actuators_state, and remove fan_setpoint_cm_h2o from ControllerStatus
+
+  // Update the outputs from the PID
+  actuators_execute(actuators_state);
+
+  // Update some status info
+  controller_status.fan_power = actuators_state.fan_power;
+  controller_status.fan_setpoint_cm_h2o = actuators_state.fan_setpoint_cm_h2o;
+
+  // Pet the watchdog
+  Hal.watchdog_handler();
+}
+
+// This function is the lower priority background loop which runs continuously
+// after some basic system init.  Pretty much everything not time critical
+// should go here.
+static void background_loop() {
+
+  // Calibrate the sensors
+  sensors.Calibrate();
 
   // Current controller status.  Updated when we receive data from the GUI, when
   // sensors read data, etc.
-  ControllerStatus controller_status = ControllerStatus_init_zero;
+  controller_status = ControllerStatus_init_zero;
```

Comment:
This is safe to do without a BlockInterrupts because startLoopTimer hasn't been called yet, correct?

Reply:
Right, there isn't a race condition there.